sequencer/src/api.rs

```diff
@@ -379,15 +379,95 @@ pub mod test_helpers {
     use tide_disco::error::ServerError;
-    use url::Url;
     use vbs::version::Version;
 
     pub const STAKE_TABLE_CAPACITY_FOR_TEST: u64 = 10;
 
-    pub struct TestNetwork<P: SequencerPersistence> {
-        pub server: SequencerContext<network::Memory, P, SequencerVersion>,
-        pub peers: Vec<SequencerContext<network::Memory, P, SequencerVersion>>,
-        pub cfg: TestConfig,
+    pub struct TestNetwork<P: PersistenceOptions, const NUM_NODES: usize> {
+        pub server: SequencerContext<network::Memory, P::Persistence, SequencerVersion>,
+        pub peers: Vec<SequencerContext<network::Memory, P::Persistence, SequencerVersion>>,
+        pub cfg: TestConfig<{ NUM_NODES }>,
+    }
+
+    pub struct TestNetworkConfig<const NUM_NODES: usize, P, C>
+    where
+        P: PersistenceOptions,
+        C: StateCatchup + 'static,
+    {
+        state: [ValidatedState; NUM_NODES],
+        persistence: [P; NUM_NODES],
+        catchup: [C; NUM_NODES],
+        network_config: TestConfig<{ NUM_NODES }>,
+        api_config: Options,
+    }
+
+    pub struct TestNetworkConfigBuilder<const NUM_NODES: usize, P, C>
+    where
+        P: PersistenceOptions,
+        C: StateCatchup + 'static,
+    {
+        state: [ValidatedState; NUM_NODES],
+        persistence: Option<[P; NUM_NODES]>,
+        catchup: Option<[C; NUM_NODES]>,
+        api_config: Option<Options>,
+        network_config: Option<TestConfig<{ NUM_NODES }>>,
+    }
+
+    impl<const NUM_NODES: usize, P, C> Default for TestNetworkConfigBuilder<{ NUM_NODES }, P, C>
+    where
+        P: PersistenceOptions,
+        C: StateCatchup + 'static,
+    {
+        fn default() -> Self {
+            Self {
+                state: std::array::from_fn::<_, { NUM_NODES }, _>(|_| ValidatedState::default()),
+                persistence: None,
+                catchup: None,
+                network_config: None,
+                api_config: None,
+            }
+        }
+    }
+
+    impl<const NUM_NODES: usize, P, C> TestNetworkConfigBuilder<{ NUM_NODES }, P, C>
+    where
+        P: PersistenceOptions,
+        C: StateCatchup + 'static,
+    {
+        pub fn states(mut self, state: [ValidatedState; NUM_NODES]) -> Self {
+            self.state = state;
+            self
+        }
+
+        pub fn persistences(mut self, persistence: [P; NUM_NODES]) -> Self {
+            self.persistence = Some(persistence);
+            self
+        }
+
+        pub fn api_config(mut self, api_config: Options) -> Self {
+            self.api_config = Some(api_config);
+            self
+        }
+
+        pub fn catchups(mut self, catchup: [C; NUM_NODES]) -> Self {
+            self.catchup = Some(catchup);
+            self
+        }
+
+        pub fn network_config(mut self, network_config: TestConfig<{ NUM_NODES }>) -> Self {
+            self.network_config = Some(network_config);
+            self
+        }
+
+        pub fn build(self) -> TestNetworkConfig<{ NUM_NODES }, P, C> {
```

Comment:
What is the point of `TestNetworkConfig`? Why can't we go directly from `TestNetworkConfigBuilder` (or call it `TestNetworkBuilder` to `TestNetwork`?)

Reply:
I think the configuration of a `TestNetwork` is too complecated and it should be more readable and understandable by this. `TestNetwork` provides the accessibility to the running network and `TestNetWorkConfig` defines the network.
As for `Builder` things, actually I do think it is a bit troublesome but builder pattern is prevalent and people can expect what is going on easily. After calling `.build()` the config is fixed and this prevents people modifying the config by `set_xxxx`(I think this is a mess).